### ISSUE TITLE
Improve configuration/composition of Prometheus metric collection

### ISSUE THREAD

**Issue by hendrikmakait** [dask/distributed#7364]
As of now, our Prometheus collectors are hard-coded into their respective submodules. This makes it hard to make changes to the metrics we expose. In particular, I see two use cases where better configuration/composability would benefit users:

1. Some of the metrics we expose are very detailed and mostly useful for debugging and development purpose. Since there is a cost to each collected metric (in terms of storage, processing time, or $ spent on Cloud services), it would be helpful to be able to configure which (sets of) metrics we want to expose for scraping.
2. Extensions or plugins may want to expose additional metrics, see the `WorkStealingMetricCollector` as an example. For these, having an easy way to (un)register collectors would be great.

**Comment by gjoseph92** [comment]
I'd suggest reframing this as just "use the Prometheus python client in the way it's [more typically used](https://github.com/prometheus/client_python#instrumenting)". In my experience with Prometheus, when you want to add a metric, you'd just do something like:

```python
# foo.py
from prometheus_client import Counter
foo_count = Counter("foo_count", "How many foos have occurred")
...
def on_new_foo(...):
    foo_count.inc()
```

```python
# bar.py
from prometheus_client import Gauge, Histogram
active_bars = Gauge("active_bars", "How many bars at the moment")
bar_latency = Histogram("bar_latency", "How long it takes to make a bar")
...
def add_bar(...):
    with bar_latency.time():
        bar = Bar(...)
    active_bars.inc()

def close_bar(...):
    active_bars.dec()
```
That's all you have to do. Some global-variable magic in the Prometheus client collects all these metric instances across the various files in an internal registry, and when you collect them in a scrape, it just reads their values.

This makes working with metrics easier as a developer because:
1. Metrics are defined in the place where they happen
1. Metrics are updated when the _change_ occurs, not when the scrape occurs.

    As a dev, you don't have to think about Prometheus's pull-based model: when the thing happens that you care about, you just update the metric. The value is just stored for you until the next scrape. Right now, we have all this code in `collect` that has to pull the current values out of the system when a scrape happens.

    For metrics where that's not easy to do (https://github.com/dask/distributed/pull/7370, https://github.com/dask/distributed/pull/7368 are good examples), we're kind of reinventing the wheel by storing the metric in an internal variable when the change occurs, so we can access that when a scrape happens. This is exactly what normal use of the Prometheus client would handle for you.
1. Because of 1. and 2., you can easily even write tests for metrics too, since you don't have to call a `collect` function to get their values.

A couple things we'll have to think about to make this change:
1. Hooking the client's collection mechanism into our Tornado HTTP server
2. (not strictly necessary) finding an asyncio-compatible prometheus client, so if multiple async `Worker`s are in the same process (aka unit tests), their metrics can be isolated. Not necessary for production, but could be handy for testing metrics.

After that, it'll just be some refactoring to define and update our current metrics in the places where they actually happen.

**Comment by gjoseph92** [comment]
> Since there is a cost to each collected metric

I wonder how much to worry about that (maybe @ntabris can weigh in). AFAIU metrics are extremely cheap (and if storage is actually expensive there's always [VictoriaMetrics](https://victoriametrics.com/)). The thing that makes metrics expensive is when they're _high-cardinality_.

> As a general guideline, try to keep the cardinality of your metrics below 10, and for metrics that exceed that, aim to limit them to a handful across your whole system.
https://prometheus.io/docs/practices/instrumentation/#do-not-overuse-labels

As far as actual CPU time to manage the metrics:

> In general, the additional resource cost of instrumentation is far outweighed by the benefits it brings to operations and development.
> 
> For code which is performance-critical or called more than 100k times a second inside a given process, you may wish to take some care as to how many metrics you update.
> 
> A Java counter takes [12-17ns](https://github.com/prometheus/client_java/blob/master/benchmark/README.md) to increment depending on contention. Other languages will have similar performance.
https://prometheus.io/docs/practices/instrumentation/#inner-loops

**Comment by crusaderky** [comment]
+1 from me. It would let us clean up a lot of secondary steps for collection, drop crick entirely, and remove ugly hacks like the local maximums that are wiped clean at every collection.

**Comment by crusaderky** [comment]
I jotted down #7411 to get a feel of how it would look like, and *I love it*.

Caveats:
- Many measures would change name - because we're using histograms instead of max and we're introducing summaries instead of _bytes+_count couples. We need a proper deprecation/versioning system.
- It would be impossible to have per-worker or per-scheduler metrics in unit tests, when there are more than one Scheduler/Worker running in the same process at once. I guess we can rig `@gen_cluster` to wipe the metrics at the beginning and end of every test.

**Comment by dchudz** [comment]
> We need a proper deprecation/versioning system.

Do we have a sense for how widely used these metrics are?

**Comment by crusaderky** [comment]
>> We need a proper deprecation/versioning system.
> 
> Do we have a sense for how widely used these metrics are?

Discussion here: https://github.com/dask/distributed/issues/7385

**Comment by dchudz** [comment]
@gjoseph92 summarizing our in-person discussion (this time on the right issue):

I think you're planning on working on this soon, with an approach that shouldn't take more than a week or so of engineering effort.

I also think you'll circulate a design/plan this with, in order to get buy-in from Florian etc.

Right?

**Comment by gjoseph92** [comment]
Yup, I'll build off of @crusaderky's draft PR and figure out a compatibility layer for when prometheus isn't installed.

**Comment by fjetter** [comment]
> I'd suggest reframing this as just "use the Prometheus python client in the way it's [more typically used](https://github.com/prometheus/client_python#instrumenting)". In my experience with Prometheus, when you want to add a metric, you'd just do something like:

FWIW I believe the reason why this path was chosen because prometheus was always a second-class dependency in distributed and a compatibility layer was not worth it.

---

About the migration itself, we're also collecting metrics that do not fit perfectly into the "use simple prometheus client" approach. At least not the way they are calculating and defining them right now. Most specifically, we started to collect digests, e.g. max tick duration, as a Gauge and we reset this counter after every collection. This is obviously already a work around to the current architecture.
If we move to prometheus client directly I believe we'll need a histogram. That's not necessarily a bad thing and might indeed turn out quite nicely. However, it is a non-trivial change for consumers and we'll need to configure buckets specifically.

When thinking of a compatibility layer, I would actually invite us to think a bit broader than purely wrapping prometheus. Often we're actually interested in spans / tracing, e.g. https://github.com/dask/distributed/issues/7217 https://github.com/dask/distributed/pull/7484
This is out of scope for this ticket but if we introduce a compat layer I would like to keep these applications in mind for a future enhancement. We're likely not locked out of it either way, just providing additional context.

**Comment by hendrikmakait** [comment]
> if we introduce a compat layer I would like to keep these applications in mind for a future enhancement

From my experience with #7484, @gjoseph92's approach is a good incremental step in the right direction. We'd have to do something very similar if we decided to move from `prometheus-client` to OTel.
